185769334 apr and caper 2024 spec

ttoomey opened the pull request:
## Description
[Update APR/CAPER to 2024 spec
](https://www.pivotaltracker.com/story/show/185769334)

* first pass through APR/CAPER 2024. Tests are not green yet
* includes datalab spec work from Elliot
* migrations add fields to hud_report_apr_clients to support new questions (needs review)
* added QuestionSheet abstraction to reduce overhead around generating questions
* added label_for helper for common vocab elements such as dkptr
* moved shared question code up into base which is not great, this class is too large
* changed common hud report mixins for 2024 per discussion with Elliot


## Type of change
- [x] New feature (adds functionality)

## Checklist before requesting review
- [ ] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [ ] I have updated the documentation (or not applicable)
- [ ] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

ttoomey committed:
185769334-apr-and-caper-2024-spec

ttoomey committed:
185769334-apr-and-caper-2024-spec

age range/label

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 10

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 4

ttoomey committed:
185769334-apr-and-caper-2024-spec

bug fix select

ttoomey committed:
185769334-apr-and-caper-2024-spec

quesion 4

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 10

ttoomey committed:
185769334-apr-and-caper-2024-spec

rubocop

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 6

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 7

ttoomey committed:
185769334-apr-and-caper-2024-spec

race_ethnicity_multi

ttoomey committed:
185769334-apr-and-caper-2024-spec

q6

ttoomey committed:
185769334-apr-and-caper-2024-spec

q6b

ttoomey commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_six.rb:
```diff
@@ -0,0 +1,906 @@
+      q6b_disabling_condition(sheet)
+    end
+
+    def q6b_veteran_answers(sheet:)
```
Hi @eanders 
Does this all seem okay? I was trying to make this a little easier to work with and q6b needed to be reworked anyway for 2024. I'm using this pattern on the q6b questions but haven't really tested it yet
* I added a light "sheet" abstraction to help reduce boilerplate
* I'm not attaching members to the "total" cell since we already have them on the row
* using "value" as an alias for cell.summary

eanders commented on app/models/concerns/hud_concerns/client.rb:
```diff
@@ -45,6 +45,16 @@ def gender_multi
 
+    def race_multi
+      ret = []
+      HudUtility2024.race_field_name_to_id.except(:RaceNone).each do |k, v|
```
I'm probably missing something, but this is a hash keyed on an integer, so I think `except(:RaceNone)` should be removed or replaced with `except(8, 9, 99)`

eanders replied:
Huh, I guess I'm wrong.

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/caper/fy2024/question_twelve.rb:
```diff
@@ -0,0 +1,20 @@
+    def run_question!
+      @report.start(QUESTION_NUMBER, QUESTION_TABLE_NUMBERS)
+
+      q12a_race
+      q12b_ethnicity
```
Pretty sure these are now only one table.  Also above on line 9

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/apr/fy2024/question_twelve.rb:
```diff
@@ -0,0 +1,20 @@
+
+module HudApr::Generators::Apr::Fy2024
+  class QuestionTwelve < HudApr::Generators::Shared::Fy2024::QuestionTwelve
+    QUESTION_TABLE_NUMBERS = ['Q12a', 'Q12b'].freeze
```
12 is now just one table.

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/base.rb:
```diff
@@ -0,0 +1,826 @@
+    # NOTE: HMIS allows for clients to report multiple races. The APR however, does not, and has a single
+    # race field. The order that the races appear in the report is encoded in the 'order' of this hash.
+    # This practice is very brittle, so we'll copy those here and hard code those relationships
+    private def races
```
Not sure if this is in use any more, but it will need to be updated with the new race categories.

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_sheet.rb:
```diff
@@ -0,0 +1,23 @@
+# utility class to dry up answers
+module HudApr::Generators::Shared::Fy2024
+  class QuestionSheet
+    attr_reader :report, :question
+    def initialize(report:, question:)
+      @report = report
+      @question = question
+    end
+
+    def update_cell_value(cell:, value:)
+      answer = @report.answer(question: question, cell: cell)
+      answer.update(summary: value)
+      answer
+    end
+
+    def update_cell_members(cell:, members:)
+      answer = @report.answer(question: question, cell: cell)
+      answer.add_members(members)
+      answer.update(summary: members.size)
+      answer
+    end
+  end
+end
```
This looks great!  Way cleaner.  Thanks! 
(let's add the boiler plate at the top.)

```
###
 # Copyright 2016 - 2023 Green River Data Analysis, LLC
 #
 # License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
 ###
```

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_six.rb:
```diff
@@ -0,0 +1,908 @@
+      @report.answer(question: table_name).update(metadata: metadata)
+
+      sheet = question_sheet(question: table_name)
+      q6b_veteran_row(sheet)
+      q6b_project_row(sheet)
+      q6b_hoh_relationship_row(sheet)
+      q6b_client_location_row(sheet)
+      q6b_disabling_condition_row(sheet)
```
do these need to return the sheet?
```suggestion
      q6b_veteran_row(sheet)
      q6b_project_row(sheet)
      q6b_hoh_relationship_row(sheet)
      q6b_client_location_row(sheet)
      q6b_disabling_condition_row(sheet)
```

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 12

ttoomey committed:
185769334-apr-and-caper-2024-spec

question 12

eanders committed:
Framework for FY2024 HUD report tests (#3629)

* Test infrastructure for FY2024 HUD reports

* Setup fixpoints to speed up tests

* WIP; debugging to try and track down why APR tests are slow

* Standardize some test db names

* Align db names for CI

* Re-adding all tests

* Remove non-APR tables from the test kit

* Re-enable all APR/CAPER tests, updated for FY2024

ttoomey committed:
Merge branch '185769334-apr-and-caper-2024-spec' of github.com:greenriver/hmis-warehouse into 185769334-apr-and-caper-2024-spec

ttoomey committed:
185769334-apr-and-caper-2024-spec

q12, 13, 14

ttoomey committed:
185769334-apr-and-caper-2024-spec

q15

ttoomey committed:
185769334-apr-and-caper-2024-spec

q16

ttoomey committed:
185769334-apr-and-caper-2024-spec

q19

ttoomey committed:
185769334-apr-and-caper-2024-spec

q20

ttoomey committed:
185769334-apr-and-caper-2024-spec

q21

ttoomey committed:
185769334-apr-and-caper-2024-spec

q22

ttoomey committed:
185769334-apr-and-caper-2024-spec

disable prior years

ttoomey committed:
185769334-apr-and-caper-2024-spec

fixes

ttoomey committed:
185769334-apr-and-caper-2024-spec

q22
safe inspect
add label helper
adjust global hud report mixins

ttoomey committed:
Merge branch 'release-90' into 185769334-apr-and-caper-2024-spec

ttoomey committed:
185769334-apr-and-caper-2024-spec

rubocop

ttoomey committed:
185769334-apr-and-caper-2024-spec

q23

ttoomey committed:
185769334-apr-and-caper-2024-spec

q23d

ttoomey committed:
185769334-apr-and-caper-2024-spec

q23e

ttoomey committed:
185769334-apr-and-caper-2024-spec

q24

ttoomey committed:
185769334-apr-and-caper-2024-spec

q25

ttoomey committed:
185769334-apr-and-caper-2024-spec

q25

ttoomey committed:
185769334-apr-and-caper-2024-spec

q26

ttoomey committed:
185769334-apr-and-caper-2024-spec

pit calc for project type 0 (q8b)

ttoomey committed:
185769334-apr-and-caper-2024-spec

schema

ttoomey committed:
185769334-apr-and-caper-2024-spec

rubocop

ttoomey committed:
185769334-apr-and-caper-2024-spec

q14b missed changes

ttoomey committed:
185769334-apr-and-caper-2024-spec

cleanp / bugfix

ttoomey committed:
185769334-apr-and-caper-2024-spec

q4

ttoomey committed:
185769334-apr-and-caper-2024-spec

bugfix

ttoomey committed:
185769334-apr-and-caper-2024-spec

fix datalab project names

ttoomey committed:
185769334-apr-and-caper-2024-spec

improve sheet builder method names

ttoomey committed:
185769334-apr-and-caper-2024-spec

dkptr label

ttoomey committed:
185769334-apr-and-caper-2024-spec

apr q24

ttoomey committed:
185769334-apr-and-caper-2024-spec

q25

ttoomey committed:
185769334-apr-and-caper-2024-spec

q26

ttoomey committed:
185769334-apr-and-caper-2024-spec

q27

ttoomey committed:
185769334-apr-and-caper-2024-spec

q27m

ttoomey committed:
185769334-apr-and-caper-2024-spec

rubocop

ttoomey committed:
185769334-apr-and-caper-2024-spec

fixes

ttoomey committed:
Merge branch 'release-90' into 185769334-apr-and-caper-2024-spec

ttoomey committed:
185769334-apr-and-caper-2024-spec

schema

ttoomey commented:
@eanders I have both the APR and CAPER running locally. I believe I have all new questions for 2024 completed. I'm planning to make another pass for review and cleanup tomorrow

eanders commented on app/models/concerns/hud_reports/clients.rb:
```diff
@@ -26,12 +26,23 @@ module HudReports::Clients
       }
     end
 
+    # override HudReports::Clients concern
+    def yes_know_dkn_clauses(column)
+      {
+        'Yes' => column.eq(1),
+        'No' => column.eq(0),
+        label_for(:dkptr) => column.in([8, 9]),
+        label_for(:data_not_collected) => column.eq(99).or(column.eq(nil)),
+        'Total' => Arel.sql('1=1'),
+      }
+    end
```
Pretty sure we can just put this in `HudReports::Clients`  I'm not seeing a functional difference.

ttoomey replied:
for sure, thanks! Fwiw I haven't completed a code review yet so I'm sure there's lingering cruft

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/base.rb:
```diff
@@ -0,0 +1,1112 @@
+      ].freeze
+    end
+
+    private def age_ranges
+      {
+        'Under 5' => a_t[:age].between(0..4),
+        '5-12' => a_t[:age].between(5..12),
+        '13-17' => a_t[:age].between(13..17),
+        '18-24' => a_t[:age].between(18..24),
+        '25-34' => a_t[:age].between(25..34),
+        '35-44' => a_t[:age].between(35..44),
+        '45-54' => a_t[:age].between(45..54),
+        '55-64' => a_t[:age].between(55..64),
+        '65+' => a_t[:age].gteq(65),
+        label_for(:dkptr) => a_t[:dob_quality].in([8, 9]).and(a_t[:dob].eq(nil)),
+        'Data Not Collected' => a_t[:dob_quality].eq(99).and(a_t[:dob].eq(nil)),
+        'Total' => Arel.sql('1=1'), # include everyone
+      }
+    end
```
Should this live in `app/models/concerns/hud_reports/ages.rb` ?  Does it not already live there?

ttoomey replied:
yeah it was a duplicate, I move it to the mixin

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/base.rb:
```diff
@@ -0,0 +1,1112 @@
+        'Man' => [3, gender_col.eq('1')],
+        'Culturally Specific Identity' => [4, gender_col.eq('2')],
+        'Transgender' => [5, gender_col.eq('5')],
+        'Non-Binary' => [6, gender_col.eq('4')],
+        'Questioning' => [7, gender_col.eq('6')],
+        'Different Identity' => [8, gender_col.eq('3')],
+
+        'Woman/Man' => [9, gender_col.eq('0,1')],
+        'Woman/Culturally Specific Identity' => [10, gender_col.eq('0,2')],
+        'Woman/Transgender' => [11, gender_col.eq('0,5')],
+        'Woman/Non-Binary' => [12, gender_col.eq('0,4')],
+        'Woman/Questioning' => [13, gender_col.eq('0,6')],
+        'Woman/Different Identity' => [14, gender_col.eq('0,3')],
+
+        'Man/Culturally Specific Identity' => [15, gender_col.eq('1,2')],
+        'Man/Transgender' => [16, gender_col.eq('1,5')],
+        'Man/Non-Binary' => [17, gender_col.eq('1,4')],
+        'Man/Questioning' => [18, gender_col.eq('1,6')],
+        'Man/Different Identity' => [19, gender_col.eq('1,3')],
+
+        'Culturally Specific Identity/Transgender' => [20, gender_col.eq('2,5')],
+        'Culturally Specific Identity/Non-Binary' => [21, gender_col.eq('2,4')],
+        'Culturally Specific Identity/Questioning' => [22, gender_col.eq('2,6')],
+        'Culturally Specific Identity/Different Identity' => [23, gender_col.eq('2,3')],
+
+        'Transgender/Non-Binary' => [24, gender_col.eq('5,4')],
+        'Transgender/Questioning' => [25, gender_col.eq('5,6')],
+        'Transgender/Different Identity' => [26, gender_col.eq('5,3')],
+
+        'Non-Binary/Questioning' => [27, gender_col.eq('4,6')],
+        'Non-Binary/Different Identity' => [28, gender_col.eq('4,3')],
+
+        'Questioning/Different Identity' => [29, gender_col.eq('6,3')],
+        # 2 or more commas
+        'More than 2 Gender Identities Selected' => [30, gender_col.matches_regexp('(\d+,){2,}')],
+        label_for(:dkptr) => [31, gender_col.in(['8', '9'])],
+        'Data Not Collected' => [32, gender_col.eq('99')],
+        'Total' => [33, Arel.sql('1=1')],
+      }.freeze
+    end
```
This should probably live in  `app/models/concerns/hud_reports/clients.rb`

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/base.rb:
```diff
@@ -0,0 +1,1112 @@
+      }.freeze
+    end
+
+    def sub_populations_by_subsidy_type(question:, members:)
```
I think I'd rather see the following methods batched up into a concern, if the aren't used in the generation of the `hud_apr_clients` table, I think it'll be cleaner to have them out of this file.

ttoomey replied:
Yeah, this is definitely bloating this class; it was quick and dirty on my part. However if we're going to fix it, I worry a little bit about adding more mixins. With such a wide model, name collisions could be an issue and mixins make us blind to that. I'll try and extract these these into new classes, composition is probably a safer path

ttoomey replied:
it was taking too long to extract these methods into classes so I went with more mixins. Ah well

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_four.rb:
```diff
@@ -0,0 +1,101 @@
+      # Column G should show the response to 2.09.1 effective as of the [report end date]. This element is a transactional data element, and only the most recent value for the report period is displayed.
+      records = project.ce_participations.filter do |record|
+        start = record.CEParticipationStatusStartDate || 100.years.ago
+        stop = record.CEParticipationStatusEndDate || 100.years.from_now
```
```suggestion
        stop = record.CEParticipationStatusEndDate || Date.current```

eanders commented on drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_four.rb:
```diff
@@ -0,0 +1,101 @@
+      records = project.ce_participations.filter do |record|
+        start = record.CEParticipationStatusStartDate || 100.years.ago
+        stop = record.CEParticipationStatusEndDate || 100.years.from_now
+        @report.start_date.between?(start, stop)
```
```suggestion
        @report.end_date.between?(start, stop)
```
Otherwise I believe this is correct.  Basically find the most-recently changed record that overlapped the report end date.

ttoomey committed:
185769334-apr-and-caper-2024-spec

pr feedback

ttoomey committed:
185769334-apr-and-caper-2024-spec

pr feedback, extract questions

eanders committed:
Q4a adjustments for tests

ttoomey committed:
185769334-apr-and-caper-2024-spec

fix incorrect question numbers
move question sheet to hud_reports namespace for re use

eanders committed:
Reported Q5a to AirTable

eanders committed:
Additional test adjustments

eanders committed:
WIP; additional fixes for failing tests

eanders committed:
WIP; additional fixes for failing tests

eanders committed:
Additional fixes for tests

eanders committed:
APR Multiple Projects first-pass test discovery complete

eanders committed:
Merge branch 'release-90' into 185769334-apr-and-caper-2024-spec

eanders committed:
Restore Q25i column

eanders committed:
Q25i fixes
